--- /Users/arthurvargas/dev/src/git/market-scout/src/patch/ibpy2_original_dispatcher.py	2024-09-25 13:03:56.000000000 -0500
+++ /Users/arthurvargas/dev/src/git/market-scout/src/patch/ibpy2_modified_dispatcher.py	2024-09-25 13:05:07.000000000 -0500
@@ -5,7 +5,7 @@
 # Defines Dispatcher class to send messages to registered listeners.
 #
 ##
-from Queue import Queue, Empty
+from queue import Queue, Empty
 
 from ib.lib import maybeName, logger
 from ib.opt import message
