--- /Users/arthurvargas/dev/src/git/market-scout/src/patch/ibpy2_original_eclient_socket.py	2024-09-26 09:56:37.000000000 -0500
+++ /Users/arthurvargas/dev/src/git/market-scout/src/patch/ibpy2_modified_eclient_socket.py	2024-09-26 09:57:00.000000000 -0500
@@ -293,10 +293,10 @@
         self.m_reader = self.createReader(self, DataInputStream(socket.getInputStream()))
         #  check server version
         self.m_serverVersion = self.m_reader.readInt()
-        print "Server Version: %d" % self.m_serverVersion
+        print("Server Version: %d" % self.m_serverVersion)
         if self.m_serverVersion >= 20:
             self.m_TwsTime = self.m_reader.readStr()
-            print "TWS Time at connection:" + self.m_TwsTime
+            print ("TWS Time at connection:" + self.m_TwsTime)
         if self.m_serverVersion < self.SERVER_VERSION:
             self.eDisconnect()
             self.m_anyWrapper.error(EClientErrors.NO_VALID_ID, EClientErrors.UPDATE_TWS.code(), EClientErrors.UPDATE_TWS.msg())
